File: lib/src/main/java/androidx/lifecycle/ViewModelHack.java
package androidx.lifecycle;

import androidx.annotation.NonNull;

/**
 * Created on 29.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */
public class ViewModelHack {

	private static final String DEFAULT_KEY = "androidx.lifecycle.ViewModelProvider.DefaultKey";

	private static @NonNull String defaultKey(Class<?> modelClass) {
		String canonicalName = modelClass.getCanonicalName();
		if (canonicalName == null) {
			throw new IllegalArgumentException("Local and anonymous classes can not be ViewModels");
		}
		return DEFAULT_KEY + ":" + canonicalName;
	}

	/**
	 * Just a way to get {@link ViewModel} without creating {@link ViewModelProvider}
	 * and {@link ViewModelProvider.Factory}.
	 */
	@SuppressWarnings("unchecked")
	public static <T extends ViewModel> T getOrNewInstance(@NonNull Class<T> modelClass,
	                                                       @NonNull ViewModelStore store) {
		String key = defaultKey(modelClass);
		ViewModel viewModel = store.get(key);

		if (modelClass.isInstance(viewModel)) {
			return (T) viewModel;
		}

		try {
			viewModel = modelClass.newInstance();
		}
		catch (IllegalAccessException e) {
			throw new RuntimeException("Cannot create an instance of " + modelClass, e);
		}
		catch (InstantiationException e) {
			throw new RuntimeException("Cannot create an instance of " + modelClass, e);
		}

		store.put(key, viewModel);

		return (T) viewModel;
	}

}

File: lib/src/main/java/androidx/lifecycle/ViewModelHack.kt
package androidx.lifecycle

/**
 * Created on 16.03.2019.
 * @author dhabensky <dhabensky@yandex.ru>
 */

fun clearViewModel(viewModel: ViewModel?) {
    viewModel?.onCleared()
}

fun putViewModel(key: String, viewModel: ViewModel, store: ViewModelStore) {
    store.put(key, viewModel)
}

fun getViewModel(key: String, store: ViewModelStore): ViewModel? {
    return store.get(key)
}


private val DEFAULT_KEY = "androidx.lifecycle.ViewModelProvider.DefaultKey"

fun defaultKey(modelClass: Class<*>): String {
    val canonicalName = modelClass.canonicalName
        ?: throw IllegalArgumentException("Local and anonymous classes can not be ViewModels")
    return "$DEFAULT_KEY:$canonicalName"
}
